bundles/org.openhab.io.homekit/src/main/java/org/openhab/io/homekit/internal/accessories/HomekitAccessoryFactory.java

```diff
@@ -12,156 +12,417 @@
+                                                                                                               // as
+                                                                                                               // well
+                    accessories
+                            .add(new SimpleEntry<>(type, HomekitCharacteristicType.valueOfTag(meta[1].trim()).get()));
+                } else {// it has no characteristic
+                    accessories.add(new SimpleEntry<>(type, EMPTY));
+                }
+            } else { // it is no accessory, so, maybe it is a characteristic
+                HomekitCharacteristicType.valueOfTag(meta[0].trim())
+                        .ifPresent(c -> accessories.add(new SimpleEntry<>(DUMMY, c)));
+            }
+        }
+        return accessories;
+    }
+
+    public static @Nullable Map<String, Object> getItemConfiguration(Item item, MetadataRegistry metadataRegistry) {
+        Metadata metadata = metadataRegistry.get(new MetadataKey(METADATA_KEY, item.getUID()));
+        return metadata != null ? metadata.getConfiguration() : null;
     }
 
     /**
-     * Given an accessory group, return the item in the group tagged as an accessory.
-     *
-     * @param taggedItem The group item containing our item, or, the accessory item.
-     * @param accessoryType The accessory type for which we're looking
-     * @return
+     * return list of HomeKit relevant groups linked to an accessory
+     * 
+     * @param item OH item
+     * @param itemRegistry item registry
+     * @param metadataRegistry metadata registry
+     * @return list of relevant group items
      */
-    private static Optional<HomekitTaggedItem> getPrimaryAccessory(HomekitTaggedItem taggedItem,
-            HomekitAccessoryType accessoryType, ItemRegistry itemRegistry) {
-        LOGGER.debug("{}: isGroup? {}, isMember? {}", taggedItem.getName(), taggedItem.isGroup(),
-                taggedItem.isMemberOfAccessoryGroup());
+    public static List<GroupItem> getAccessoryGroups(Item item, ItemRegistry itemRegistry,
+            MetadataRegistry metadataRegistry) {
+        return item.getGroupNames().stream().flatMap(name -> {
+            Item groupItem = itemRegistry.get(name);
+            if (groupItem instanceof GroupItem) {
+                return Stream.of((GroupItem) groupItem);
+            } else {
+                return Stream.empty();
+            }
+        }).filter(groupItem -> !getAccessoryTypes(groupItem, metadataRegistry).isEmpty()).collect(Collectors.toList());
+    }
+
+    /**
+     * collect all mandatory characteristics for a given tagged item, e.g. collect all mandatory HomeKit items from a
+     * GroupItem
+     * 
+     * @param taggedItem HomeKit tagged item
+     * @param metadataRegistry meta data registry
+     * @return list of mandatory
+     */
+    private static @NonNull List<HomekitTaggedItem> getMandatoryCharacteristics(HomekitTaggedItem taggedItem,
```

Comment:
This NonNull  annotation it redundant.
```suggestion
    private static List<HomekitTaggedItem> getMandatoryCharacteristics(HomekitTaggedItem taggedItem,
```